include/swift/AST/GenericSignatureBuilder.h

```diff
@@ -940,16 +940,20 @@ class GenericSignatureBuilder::RequirementSource final
 
   /// Whether there is a trailing written requirement location.
   const bool hasTrailingWrittenRequirementLoc;
-
 public:
   /// Whether a protocol requirement came from the requirement signature.
   const bool usesRequirementSignature;
 
 private:
   /// The actual storage, described by \c storageKind.
   union {
-    /// The type to which a requirement applies.
-    TypeBase *type;
+    struct {
+      /// The type to which a requirement applies.
+      TypeBase *type;
+      /// Whether this \b requirement source is for a synthetic
+      /// requirement. Can only be \c true for an abstract source.
+      bool syntheticRequirement;
```

Comment:
Rather than having this be an extra bit of state that's only valid for certain kinds, I think we should extend  `GenericSignatureBuilder::RequirementSource::Kind` with a new kind that acts like `Explicit` (which is what `forAbstract` maps to) everywhere except for ordering. I'd recommend calling it `CovariantSelf`, since that's directly what it should be.

Reply:
Hi Doug, thanks a lot for taking a look, I perfectly understand the busyness. 

It is a bit unusual for `CovariantSelf` to be a special case, but that definitely sounds like a cleaner solution. Agreed!